Implement iterators over rows

fizyk20 opened the pull request:
This implements methods on `ImageBuffer` that enable iterating over rows (which are themselves iterators over pixels in the rows).

I needed this for my project, but maybe it will prove useful for someone else, too ;)

HeroicKatora commented on src/buffer.rs:
```diff
@@ -355,6 +529,15 @@ where
+        Rows {
+            chunks: self
+                .data
+                .chunks((<P as Pixel>::channel_count() as u32 * self.width) as usize),
```
This multiplication needs to happen with type `usize`. Otherwise the resulting chunks will be too small on some images (Note that `u64` is not necessary as the complete image can be addressed).

HeroicKatora commented on src/buffer.rs:
```diff
@@ -466,6 +660,15 @@ where
+        RowsMut {
+            chunks: self
+                .data
+                .chunks_mut((<P as Pixel>::channel_count() as u32 * self.width) as usize),
```
Same as the comment above, cast both to `usize` before multiplication.

HeroicKatora commented on src/buffer.rs:
```diff
@@ -253,6 +339,50 @@ where
+    fn next(&mut self) -> Option<(u32, EnumeratePixels<'a, P>)> {
+        let y = self.y;
+        self.y += 1;
+        match self.rows.next() {
```
Stylistic choice, I prefer map here as it makes clear that no other kind of control flow is introduced.

HeroicKatora commented on src/buffer.rs:
```diff
@@ -479,6 +682,18 @@ where
+            width,
+        }
+    }
+
```
Can this get some tests? E.g. the multiplication bug can be tested not to occur with an RGB image of size (2^31, 1).

fizyk20 replied:
Hmm... Won't this work only on 64-bit architectures, though, so that the test would have to only be run on such? And won't such test allocate multiple GBs of memory (maybe that's not a problem, though)?

HeroicKatora replied:
Agh, you are correct. This could be faked though with a test-only implementation of `Pixel` and `Subpixel` that are both zero-sized types, thus not requiring any allocation in their `Vec<_>` etc. The image itself would be just small enough to also not panic on 32-bit (I think?) and there not do anything useful but also not fail?

But overall, yeah, this is pretty overkill not worth it. I initially thought it was easier.

fizyk20 commented:
Sorry it took me so long to get back to this! I addressed all the comments except the one about the tests - I need some clarification there (commented above).

HeroicKatora commented:
Since we are currently about to transition the licensing, I'd need you to accept that your contribution can be licensed under those which we want to adopt (see https://github.com/image-rs/image/issues/503 for a full explanation). If you accept, please respond with this text:

> I license past and future contributions under the dual MIT/Apache-2.0 license, allowing licensees to chose either at their option.

fizyk20 commented:
Sure :+1:

I license past and future contributions under the dual MIT/Apache-2.0 license, allowing licensees to chose either at their option.

fizyk20 commented:
Thanks, but I actually rarely contribute to other projects, so I'm not sure if it's worth it ;) Maybe I'll just ping you the next time I have something to add or something like that :)

HeroicKatora commented:
Sure, that's fine :+1:
